File: 1360-maximum-length-of-a-concatenated-string-with-unique-characters/maximum-length-of-a-concatenated-string-with-unique-characters.kt
class Solution {
    
  fun maxLength(arr: List<String>): Int {
    val sets = arr.filter { it.toSet().size == it.length }
    fun dfs(i: Int, s: Set<Char>): Int = if (i == sets.size) 0
      else max(
        if (sets[i].any { it in s }) 0 else
        sets[i].length + dfs(i + 1, s + sets[i].toSet()),
        dfs(i + 1, s)
      )
    return dfs(0, setOf())
  }
}

File: 1360-maximum-length-of-a-concatenated-string-with-unique-characters/maximum-length-of-a-concatenated-string-with-unique-characters.swift
class Solution {
    func maxLength(_ arr: [String]) -> Int {
        // Filter out strings with repeated characters as we can never use them.
        let sets: [Set<Character>] = arr.compactMap {
            let characterSet = Set($0)
            return $0.count == characterSet.count ? characterSet : nil
        }

        func backtrack(_ start: Int, _ seen: Set<Character>) -> Int {
            var result = 0
            for i in start..<sets.count where seen.intersection(sets[i]).isEmpty {
                result = max(result, sets[i].count + backtrack(i + 1, seen.union(sets[i])))
            }
            return result
        }

        return backtrack(0, [])
    }
}